cpp/src/arrow/type.cc

```diff
@@ -43,6 +43,49 @@
 #include "arrow/visitor_inline.h"
 
 namespace arrow {
+
+constexpr Type::type NullType::type_id;
```

Comment:
This is ensuring that `/\w+Type::type_id/` is a defined symbol in `libarrow.so`. Previously these have only been used within a translation unit which links into `libarrow.so`, but now I'm using them in filter.cc which gets linked to `libarrow_dataset.so`. If you remove these you'll find the linker emits an error because they're undefined in `libarrow.so`

Reply:
This seems like a quirk due to the use of `ARROW_EXPORT`, would I be correct that this issue would not occur without the export/import business?